File: backend/app/utils/formatters.py
"""
Formatting Utilities
"""
from typing import Any
from datetime import datetime
import json


def format_currency_inr(amount: float) -> str:
    """Format amount as Indian Rupees"""
    if amount >= 10000000:  # 1 crore
        return f"₹{amount/10000000:.2f} Cr"
    elif amount >= 100000:  # 1 lakh
        return f"₹{amount/100000:.2f} L"
    elif amount >= 1000:
        return f"₹{amount/1000:.2f} K"
    else:
        return f"₹{amount:.2f}"


def format_percentage(value: float, decimals: int = 2) -> str:
    """Format value as percentage"""
    return f"{value:.{decimals}f}%"


def format_number(value: float, decimals: int = 2) -> str:
    """Format number with thousand separators"""
    return f"{value:,.{decimals}f}"


def format_timestamp(dt: datetime) -> str:
    """Format datetime as ISO string"""
    return dt.isoformat()


def truncate_text(text: str, max_length: int = 100) -> str:
    """Truncate text to max length"""
    if len(text) <= max_length:
        return text
    return text[:max_length - 3] + "..."


def json_serializer(obj: Any) -> str:
    """JSON serializer for objects not serializable by default"""
    if isinstance(obj, datetime):
        return obj.isoformat()
    raise TypeError(f"Type {type(obj)} not serializable")

File: frontend/src/utils/formatters.js
export function formatCurrency(amount, decimals = 2) {
  if (amount >= 10000000) {
    return `₹${(amount / 10000000).toFixed(decimals)} Cr`
  } else if (amount >= 100000) {
    return `₹${(amount / 100000).toFixed(decimals)} L`
  } else if (amount >= 1000) {
    return `₹${(amount / 1000).toFixed(decimals)} K`
  } else {
    return `₹${amount.toFixed(decimals)}`
  }
}

export function formatPercentage(value, decimals = 2) {
  const sign = value >= 0 ? '+' : ''
  return `${sign}${value.toFixed(decimals)}%`
}

export function formatNumber(value, decimals = 2) {
  return new Intl.NumberFormat('en-IN', {
    minimumFractionDigits: decimals,
    maximumFractionDigits: decimals
  }).format(value)
}

export function formatVolume(volume) {
  if (volume >= 10000000) {
    return `${(volume / 10000000).toFixed(2)} Cr`
  } else if (volume >= 1000000) {
    return `${(volume / 1000000).toFixed(2)} M`
  } else if (volume >= 1000) {
    return `${(volume / 1000).toFixed(2)} K`
  } else {
    return volume.toString()
  }
}

export function truncateText(text, maxLength = 50) {
  if (text.length <= maxLength) return text
  return text.substring(0, maxLength) + '...'
}
